llvm/tools/llvm-profgen/ProfiledBinary.h

```diff
@@ -396,19 +402,31 @@ class ProfiledBinary {
 
   StringRef getPath() const { return Path; }
   StringRef getName() const { return llvm::sys::path::filename(Path); }
-  uint64_t getBaseAddress() const { return BaseAddress; }
-  void setBaseAddress(uint64_t Address) { BaseAddress = Address; }
+  uint64_t getPIDBaseAddress(int32_t PID) const {
+    auto Pos = BaseAddressByPID.find(PID);
+    if (Pos == BaseAddressByPID.end()) {
+      // Use preferred address as the default base address.
+      return getPreferredBaseAddress();
+    }
+
+    return Pos->second;
+  }
+  void setPIDBaseAddress(int32_t PID, uint64_t Address) {
+    BaseAddressByPID[PID] = Address;
+  }
 
   bool isCOFF() const { return IsCOFF; }
 
   // Canonicalize to use preferred load address as base address.
-  uint64_t canonicalizeVirtualAddress(uint64_t Address) {
-    return Address - BaseAddress + getPreferredBaseAddress();
+  uint64_t canonicalizeVirtualAddress(int64_t PID, uint64_t Address) {
```

Comment:
Use the same int32_t type for all PIDs?

Reply:
There's already some inconsistency within profgen here, but I will change this to an int32_t so it's at least consistent within this PR